lib/puppet/util/json.rb

```diff
@@ -0,0 +1,72 @@
+module Puppet::Util
+  module Json
+    class ParseError < StandardError
+      attr_reader :cause, :data
+
+      def self.build(original_exception, data)
+        new(original_exception.message).tap do |exception|
+          exception.instance_eval do
+            @cause = original_exception
+            set_backtrace original_exception.backtrace
+            @data = data
+          end
+        end
+      end
+    end
+
+    begin
+      require 'multi_json'
+      # Force backend detection before attempting to use the library
+      # or load any other JSON libraries
+      MultiJson.default_adapter
+
+      # Preserve core type monkey-patching done by the built-in JSON gem
+      require 'json'
+    rescue LoadError
+      require 'json'
+    end
+
+    # These methods do similar processing to the fallback implemented by MultiJson
+    # when using the built-in JSON backend, to ensure consistent behavior
+    # whether or not Puppet::Util::Json can be loaded.
+    def self.load(string, options = {})
+      if defined? MultiJson
+        begin
+          MultiJson.load(string, options)
+        rescue MultiJson::ParseError => e
+          raise Puppet::Util::Json::ParseError.build(e, string)
+        end
+      else
+        begin
+          string = string.read if string.respond_to?(:read)
+
+          if string.respond_to?(:force_encoding)
+            string = string.dup.force_encoding(::Encoding::ASCII_8BIT)
```

Comment:
Did we ever figure out why force_encoding was important?
Also, force_encoding was added to String in Ruby 1.9. We don't support 1.8 right? Will this check always be true?

Reply:
Yes it will always be true, because we do not support 1.8. I'm not sure we ever figured out why it's needed. I'd be fine removing it, so long as we understand that this WILL happen regardless when we go through MultiJson.

Reply:
It looks like they implemented "fragment loading" by interpolating the JSON string into another string, wrapping it in brackets. That failed, I think we recently had the same problem in puppet.
See https://github.com/intridea/multi_json/pull/76/files
Note this is still how they manage "fragment parsing" for okjson https://github.com/intridea/multi_json/blob/master/lib/multi_json/adapters/ok_json.rb#L12

They had complaints from the community about both that causing issues with UTF-8 strings and that it removed quirks mode by default. They then updated to both force the encoding, and stop interpolating the json within a string instead using quirks mode. I have a feeling that one was unnecessary with the other change, but I don't have a great way to to verify that is what happened.

Reply:
I'm going to remove this whole block from our code, persisting it seems a little cargo-culty.